<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Err Handling</title>
</head>
<body>
    <script>
        class MathUtil {
            static sum(...numbers){

                if(numbers.length === 0){
                    throw new Error("Total parameter harus lebih dari 0")
                }

                let total = 0
                for(const number of numbers){
                    total += number
                }
                return total
            }
        }

        // error handling try catch finally
        try{
            console.info(MathUtil.sum())
            console.info("Kode block Try Akan Berhenti!")
        }catch(error){
            console.info(`Terjadi error : ${error.message}`)
            //finaly akan selalu dieksekusi terus mau itu err atau engga
        }finally{
            console.info("Program selesai")
        }
        //end

        console.info("Kode program tidak berhenti")

        // Trick memakai try finally tanpa memakai catch
        class Counter {
            #counter = 1

            next(){
                try{
                    return this.#counter
                }finally{
                    this.#counter++
                }
            }
        }
        //end

        const counter = new Counter()
        console.info(counter.next())
        console.info(counter.next())
        console.info(counter.next())
        console.info(counter.next())
        console.info(counter.next())
    </script>
</body>
</html>